test/browser/components.test.js

```diff
@@ -2572,6 +2572,35 @@ describe('Components', () => {
 			expect(scratch.innerHTML).to.equal('baz');
 			expect(callbackState).to.deep.equal({ foo: 'baz' });
 		});
+
+		// #2716
+		it('should work with readonly state', () => {
+			let update;
+			class Foo extends Component {
+				constructor(props) {
+					super(props);
+					this.state = { foo: 'bar' };
+					update = () =>
+						this.setState(prev => {
```

Comment:
Does it also work when passing an object with read only props to `setState`? I.e. `const foo = {}; Object.defineProperty(....); setState(foo)`

Reply:
Just checked and that scenario is unaffected, because we create a new state copy immediately inside `setState` 👍